File: leetcode/59. Spiral Matrix II.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        max_x = n
        max_y = n
        min_x = 0
        min_y = 0
        ans = [[0 for _ in range(n)] for _ in range(n)]
        num = 1
        while (min_x < max_x and min_y < max_y):
            for i in range(min_y, max_y):
                ans[min_x][i] = num
                num += 1
            min_x += 1
            for i in range(min_x, max_x):
                ans[i][max_y-1] = num
                num += 1
            max_y -= 1
            if min_x < max_x:
                for i in range(max_y-1, min_y-1, -1):
                    ans[max_x-1][i] = num
                    num += 1
                max_x -= 1
            if min_y < max_y:
                for i in range(max_x-1, min_x-1, -1):
                    ans[i][min_y] = num
                    num += 1
                min_y += 1
        return ans

File: leetcode/59. Spiral Matrix II.dart
class Solution {
  List<List<int>> generateMatrix(int n) {
    var max_x = n, min_x = 0, max_y = n, min_y = 0, num = 1;
    var ans = List.generate(n, (_) => List.filled(n, 0));
    while (min_x < max_x && min_y < max_y) {
      for (int i = min_y; i < max_y; i++) {
        ans[min_x][i] = num;
        num++;
      }
      min_x++;
      for (int i = min_x; i < max_x; i++) {
        ans[i][max_y - 1] = num;
        num++;
      }
      max_y--;
      if (min_x < max_x) {
        for (int i = max_y - 1; i >= min_y; i--) {
          ans[max_x - 1][i] = num;
          num++;
        }
        max_x--;
      }
      if (min_y < max_y) {
        for (int i = max_x - 1; i >= min_x; i--) {
          ans[i][min_y] = num;
          num++;
        }
        min_y++;
      }
    }
    return ans;
  }
}
